src/node/operator_event.rs

```diff
@@ -89,77 +101,88 @@ impl PartialEq for OperatorEvent {
-            }
-            // Prioritize writes.
-            (true, false) => Ordering::Greater,
-            (false, true) => Ordering::Less,
-            (false, false) => x.priority.cmp(&y.priority),
-        }
-    }
-}
-
 /// Ordering used in the lattice where `self < other` implies `self` *precedes* other.
 impl Ord for OperatorEvent {
     fn cmp(&self, other: &OperatorEvent) -> Ordering {
-        match (self.is_watermark_callback, other.is_watermark_callback) {
-            (true, true) => {
-                // Both of the events are watermarks, so the watermark with the lower timestamp
-                // should run first. Ties are broken by first by dependencies to ensure
-                // that writes to state occur before reads, and then by priority where a smaller
-                // number is higher priority.
-                match self.timestamp.cmp(&other.timestamp) {
-                    Ordering::Equal => match resolve_access_conflicts(&self, other) {
-                        // Prioritize other.
-                        Ordering::Equal => Ordering::Greater,
-                        ord => ord,
-                    },
-                    ord => ord,
+        match self.operator_type {
+            OperatorType::ReadOnly => {
+                match (self.is_watermark_callback, other.is_watermark_callback) {
+                    (true, true) => {
+                        // Both of the events are watermarks, so the watermark with the lower
+                        // timestamp should run first.
+                        self.timestamp.cmp(&other.timestamp)
+                    }
+                    (true, false) => {
+                        // `self` is a watermark, and `other` is a normal message callback.
+                        // TODO: can compare dependencies to increase parallelism and order on a
+                        // more fine-grained level.
+                        match self.timestamp.cmp(&other.timestamp) {
+                            Ordering::Greater => {
+                                // `self` timestamp is greater than `other`, execute `other` first.
+                                Ordering::Greater
+                            }
+                            Ordering::Equal => {
+                                // `self` timestamp is equal to `other`, execute `other` first.
+                                Ordering::Greater
+                            }
+                            Ordering::Less => {
+                                // `self` timestamp is less than `other`, run them in any order.
+                                // Assume state is time-versioned, so dependency issues should not
+                                // arise.
+                                Ordering::Equal
+                            }
+                        }
+                    }
+                    (false, true) => other.cmp(self).reverse(),
```

Comment:
Should this compare the timestamps? other.timestamp.cmp(self.timestamp).reverse()?

Reply:
Nope, because we don't really follow timestamp order here. Specifically, we arbitrarily break dependencies if the timestamps are equal, and choose to execute either the event in the lattice or the one being inserted.